test/test_s3.py

```diff
@@ -0,0 +1,134 @@
+
+
+def s3_client_new(secure, region, part_size=0):
+
+    event_loop_group = EventLoopGroup()
+    host_resolver = DefaultHostResolver(event_loop_group)
+    bootstrap = ClientBootstrap(event_loop_group, host_resolver)
+    credential_provider = AwsCredentialsProvider.new_default_chain(bootstrap)
+    tls_option = None
+    if secure:
+        opt = TlsContextOptions()
+        ctx = ClientTlsContext(opt)
+        tls_option = TlsConnectionOptions(ctx)
+
+    s3_client = S3Client(
+        bootstrap=bootstrap,
+        region=region,
+        credential_provider=credential_provider,
+        tls_connection_options=tls_option,
+        part_size=part_size)
+
+    return s3_client
+
+
+class S3ClientTest(NativeResourceTest):
+    region = "us-west-2"
+    timeout = 10  # seconds
+
+    def test_sanity(self):
+        s3_client = s3_client_new(False, self.region)
+        self.assertIsNotNone(s3_client)
+
+    def test_sanity_secure(self):
+        s3_client = s3_client_new(True, self.region)
+        self.assertIsNotNone(s3_client)
+
+    def test_wait_shutdown(self):
+        s3_client = s3_client_new(False, self.region)
+        self.assertIsNotNone(s3_client)
+
+        shutdown_event = s3_client.shutdown_event
+        del s3_client
+        self.assertTrue(shutdown_event.wait(self.timeout))
+
+
+class S3RequestTest(NativeResourceTest):
+    get_test_object_path = "/get_object_test_10MB.txt"
+    put_test_object_path = "/put_object_test_py_10MB.txt"
+    region = "us-west-2"
+    bucket_name = "aws-crt-canary-bucket"
+    timeout = 100  # seconds
+    num_threads = 0
+
+    response_headers = None
+    response_status_code = None
+    received_body_len = 0
+
+    put_body_stream = None
+
+    def _build_endpoint_string(self, region, bucket_name):
+        return bucket_name + ".s3." + region + ".amazonaws.com"
+
+    def _get_object_request(self):
+        headers = HttpHeaders([("host", self._build_endpoint_string(self.region, self.bucket_name))])
+        request = HttpRequest("GET", self.get_test_object_path, headers)
+        return request
+
+    def _put_object_request(self, file_name):
+        self.put_body_stream = open(file_name, "r+b")
+        file_stats = os.stat(file_name)
+        data_len = file_stats.st_size
+        headers = HttpHeaders([("host", self._build_endpoint_string(self.region, self.bucket_name)),
+                               ("Content-Type", "text/plain"), ("Content-Length", str(data_len))])
+        request = HttpRequest("PUT", self.put_test_object_path, headers, self.put_body_stream)
+        return request
+
+    def _on_request_headers(self, status_code, headers, **kargs):
+        self.response_status_code = status_code
+        self.assertIsNotNone(headers, "headers are none")
+        self.response_headers = headers
+
+    def _on_request_body(self, chunk, offset, **kargs):
+        self.assertIsNotNone(chunk, "the body chunk is none")
+        self.received_body_len = self.received_body_len + len(chunk)
```

Comment:
having an assert here isn't going to do anything, these assert functions just raise exceptions
But this is on a callback from C on another thread
and as I mentioned earlier, exceptions on these callbacks don't do anything to stop the operation. PyErr_WriteUnraisable() will write something to stderr, but everything is going to keep charging along

the pattern I've been using for detecting errors in void callbacks is to do like:
```
    if chunk is None:
        self.callback_error = "the body chunk is none"
```

then at the end of the test on the main thread, after all the callbacks have fired, I check if any of them reported an error, like: `self.assertIsNone(self.callback_error)`